pkg/core/interop/contract/call.go

```diff
@@ -69,6 +69,12 @@ func Call(ic *interop.Context) error {
 		return fmt.Errorf("method not found: %s/%d", method, len(args))
 	}
 	hasReturn := md.ReturnType != smartcontract.VoidType
+
+	ic.InvocationCalls = append(ic.InvocationCalls, state.ContractInvocation{
+		Hash:   u,
+		Method: method,
+		Params: stackitem.NewArray(args),
```

Comment:
Regarding size restrictions: let's summarize what have been discussed in https://github.com/CityOfZion/dora-server-NeoN3/pull/220#issuecomment-2325214725. Restriction on the number of *recorded* invocations or the number of arguments per each *recorded* invocation (or per all *recorded* invocations) is needed. I'd suggest to restrict both:
* The overall number of invocations is restricted in depth by the invocation stack size, but it's not restricted in length (only by executing GAS cost restriction). This problem should be solved the same way as restriction of the number of Notifications (https://github.com/nspcc-dev/neo-go/issues/3490). Currently in NeoGo node it's not restricted, but it may be restricted in C# node (which means we have a bug in NeoGo) or it may not be restricted at all (which means we have a cross-implementation bug). So we firstly need to solve https://github.com/nspcc-dev/neo-go/issues/3490 and after that port this solution to invocations restriction.
* The number of arguments per every recordable invocation may be restricted using the same approach as for Notifications: serialize it and check the resulting size: https://github.com/nspcc-dev/neo-go/blob/d47fe392fb144cc3662b0e5770e253f8d5207f27/pkg/core/interop/runtime/engine.go#L110
If serialization fails (due to recursive structures presence) or arguments size exceeds the limit, then this invocation wan't be stored in the database.

These size restrictions should be well-documented and it should be noted that invocation tracking system may be missing some invocations and users can't rely on the recorded invocations completely.

Reply:
1. While I agree it should be restricted, I don't think this point should be blocking  (just like it is not for notifications). 
2. This seems to count the total amount of bytes of all stack items. I don't understand  how that translates to parameter count of the method called. I can see it used to ensure it doesn't exceed a maximum total item count in the `Array` though.

I don't think we should throw away the complete invocation record because of a parameter violation. If it's recorded on chain then it was apparently still a valid invocation, regardless if we want to store all of the invocation details. Let's say it exceeds one of the limits, we could still store the contracthash, method, maybe parameter count and skip the actual parameters. Perhaps also adding a `is_valid` field to the json output that can be used as indicator how much of the information can be trusted.

i.e.
Valid
```json
"invocations": [
    {
        "contract_hash": "0x49cf4e5378ffcd4dec034fd98a174c5491e395e2",
        "method": "designateAsRole",
        "is_valid": true
        "parameters": {
            "type": "Array",
            "value": [
                {
                    "type": "Integer",
                    "value": "4"
                },
            ]
        }
    }
]
```
Invalid
```json
"invocations": [
    {
        "contract_hash": "0x49cf4e5378ffcd4dec034fd98a174c5491e395e2",
        "method": "designateAsRole",
        "is_valid": false
        "parameters": null
    }
]
```

Reply:
>  I don't think this point should be blocking 

Agree, let's then leave this restriction to #3490 and finalize Invocation logs without it.

> This seems to count the total amount of bytes of all stack items. I don't understand how that translates to parameter count of the method called.

I don't think that we need to stick to the *number* of parameters restriction, because the only thing that bothers us during Invocations collection is the resulting size of the serialized Invocations structure. We don't want it to be large and to take a lot of disc space. `Serialize` lets us ensure that serialized parameters do not occupy a lot of space and it also cares about the overall number of serialized elements (`SerializationContext.limit` is responsible for that). Thus I consider `Serialize` to be a perfect candidate for the arguments size restriction.

Also, we may cache the result of `Serialize` once, and then reuse it while storing Invocations on disc.

Reply:
I believe Roman was also concerned about the processing time, but I'll let him comment to see if he agrees using `serialize` is sufficient.

Reply:
Processing time is important, but we need to serialize arguments anyway to store Invocation logs, and if we reuse the result of Serialize, then processing time doesn't increase a lot.

And of course, the node should have a setting to disable Invocation logs, because it's a resource-demanding feature. Most of the public nodes likely won't have this feature enabled.

Reply:
> And of course, the node should have a setting to disable Invocation logs, because it's a resource-demanding feature. Most of the public nodes likely won't have this feature enabled.

Yeah that's fine. I have making it configurable on the to do list. In what section should I put the option? `ApplicationConfiguration.RPC` or `ApplicationConfiguration` (because it also affects the indexing behaviour)?

Reply:
To me it's more an `ApplicationConfiguration`-level setting, in particular, I'd place it into `config.Ledger` structure. You're right in that it affects the database behaviour, and it's possible that some other node services should be able to reach this setting in future.

Reply:
`ApplicationConfiguration` is OK, ideally we should make DBs compatible with/without this option.

Side note: serialization is also kinda a snapshot of items, so `DeepCopy` can even be avoided if we're checking the size via serialization (for notifications copying is important since this data can be reused in the same context for `System.Runtime.GetNotifications`).